### ISSUE TITLE
Vast Memory Consumption on Ubuntu 20.04

### ISSUE THREAD

**Issue by cherrywoods** [JabRef/jabref#8416]
### JabRef version

5.4 (latest release)

### Operating system

GNU / Linux

### Details on version and operating system

Ubuntu 20.04.3 with Gnome Shell 3.36.8 on Wayland

### Checked with the latest development build

- [ ] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

When opening JabRef and loading a library or creating a new one, memory consumption starts to increase quickly and continues until Memory and Swap are filled up. The first time this occurred to me, I had to force restart my PC. I think the second time the operating system force closed JabRef, at least it just closed once Memory and Swap were entirely filled. 

System Monitor does not associate the huge memory consumption (+6 GB) with JabRef. On the processes tab it lists JabRef as consuming 1.3 GB of memory. The remaining additional memory consumption doesn't show up in the process list at all. 

I am only having such memory issues when using JabRef. Every other program I use works like a charm (e.g. Firefox, PyCharm, Thunderbird, Texmaker, Spotify and others).

## Steps to reproduce
1. Open JabRef
2. Open any Library

Second option:
1. Open JabRef
2. Create a new library
3. Add entries (one is sufficient)

### Appendix

The output that the most recent development build of jabref produced:

```
Jan 12, 2022 8:02:40 AM com.sun.javafx.application.PlatformImpl startup
WARNING: Unsupported JavaFX configuration: classes were loaded from 'module org.jabref.merged.module', isAutomatic: false, isOpen: true
2022-01-12 08:02:40 [JavaFX Application Thread] org.jabref.logic.net.URLDownload.bypassSSLVerification()
WARN: Fix SSL exceptions by accepting ALL certificates
2022-01-12 08:02:42 [JavaFX Application Thread] org.jabref.gui.theme.ThemeManager.updateThemeSettings()
INFO: Not updating theme because it hasn't changed
2022-01-12 08:02:42 [JavaFX Application Thread] org.jabref.gui.theme.ThemeManager.updateThemeSettings()
INFO: Theme set to Theme{type=EMBEDDED, name='Dark.css'} with base css StyleSheet{jrt:/org.jabref/org/jabref/gui/Base.css}

(JabRef:8185): Gdk-WARNING **: 08:02:43.656: XSetErrorHandler() called with a GDK error trap pushed. Don't do that.
2022-01-12 08:02:55 [JavaFX Application Thread] org.jabref.model.database.BibDatabaseContext.getFulltextIndexPath()
INFO: Index path for /home/PATH/ANONYMISED/library.bib is /home/ANONYMISED/.local/share/JabRef/0.5a
2022-01-12 08:02:55 [pool-2-thread-1] org.jabref.gui.JabRefDialogService.notify()
INFO: Opening: '/home/PATH/ANONYMISED/library.bib'
2022-01-12 08:05:43 [JavaFX Application Thread] org.jabref.gui.JabRefDialogService.notify()
INFO: Saving library...
2022-01-12 08:05:44 [JavaFX Application Thread] org.jabref.gui.JabRefDialogService.notify()
INFO: Library saved
2022-01-12 08:05:58 [JavaFX Application Thread] org.jabref.gui.JabRefDialogService.notify()
INFO: Saving library...
2022-01-12 08:05:58 [JavaFX Application Thread] org.jabref.gui.JabRefDialogService.notify()
INFO: Library saved
```

**Comment by ThiloteE** [comment]
Seems related to interaction between JavaFX and GTK3:

- An update of GTK to version 4 or downgrade to version 2 would be worth a try. Same goes for updating JDK.

   https://stackoverflow.com/questions/55446534/how-to-fix-java22494-gdk-warning


- This forum post also tells an interesting story: http://www.babelsoft.net/forum/viewtopic.php?p=27776

   ```
   I finally understand what's happening. And it's indeed related to GTK3/GDK3.
   
   Negatron uses OpenJFX as its main graphical library. And OpenJFX 11+ has migrated from GTK2 to GTK3 but, with GTK3, a lot    of Linux distros have changed the graphical backend from X11 to Wayland. So, OpenJFX 11 couldn't work with those Linux distros using Wayland because OpenJFX 11 doesn't have a compatibility layer with Wayland, only with X11. To avoid having to develop a full compatibility layer with Wayland, OpenJFX 12+ simply added a hack to force GTK3 to work with X11.

   The thing is Negatron uses VLC to play video files. And for this to happen, the library to embed VLC within Negatron (VLCj) uses AWT, another Java graphical library. AWT can work with X11 and Wayland without issues. So, I think that AWT overrides the hack used by OpenJFX and forces the use of the newest Wayland graphical backend, screwing up OpenJFX along the way...

   So, I have 3 solutions at hand, sorted by increasing complexity and also from the worst to the best solution:
   1/ completely rewrite the file chooser dialog box in pure Java, thus killing OS integration for those common dialogues in Linux,    and hope that this GTK3 issue doesn't screw anything else up.
   2/ rewrite a part of VLCj so that it doesn't use AWT anymore.
   3/ write the Wayland compatibility layer that nobody at OpenJFX wants to write...
   ```

**Comment by cherrywoods** [comment]
Upgrading JDK to JDK 17 changed the issue, but did not entirely fix it. Instead of continuously increasing memory usage once I open a bibliography, now when I click on entries, the memory usage goes up by several MB. It doesn't decrease again when I open another entry. 

Therefore I could open three entries before nearly running out of RAM. 

The new issue is present in the latest development version, but appears not to be present in the version available via apt (that is 3.8.2). The issue is present in the latest version available from snap (5.4.6), but here the issue is more severe in the sense that opening three entries also filled up all my swap.

**Comment by ThiloteE** [comment]
JavaFX does not work with Wayland as of 2018-12 (https://bugs.openjdk.java.net/browse/JDK-8215364). The Fix employed by JavaFX is this one:

+    //Set the gtk backend to x11 on all the systems
+    `putenv("GDK_BACKEND=x11");`

This means, JavaFX forces your system to use the X11 Window System.

Now, the question arises, what will happen, if you don't have the X11 Window System installed on your system?
Do you have it installed? If not, this might have higher chances of fixing the problem than upgrading to GDK 4+.

If you are afraid of complications, you could try it on a VM first. E.g. https://www.virtualbox.org/

**Comment by ThiloteE** [comment]
Jabref v.3.8 probably works, because it was coded in Java, not in JavaFX.

**Comment by kaustubhmote** [comment]
I can confirm that the behavior is different in X11 and Wayland, although issue remains similar. I am attaching a graph of the virtual memory usage for JabRef on these two with an identical database (~4k entries). This is with v5.5 on Ubuntu 20.04 (installed via deb)

![analysis](https://user-images.githubusercontent.com/7735086/150634416-3a101628-5f55-41a7-8f8d-4650829ceee9.png)

The initial jump to ~10 GB in both the systems comes just after starting JabRef. I let it sit for a few seconds to get the baseline, and as soon as I do anything in the main table (in this case, open an entry in the preview by double-clicking), the usage jumps immediately to ~120GB on both. The rest of the time, I am adding new entries by importing pdfs. In X11 the memory usage remains constant for a while, and everything seems alright, but there is a sudden jump in usage, which hangs JabRef and I have to kill the process. In Wayland, there is a gradual increase in memory consumption and the process gets killed by the OS automatically after a certain threshold.
 
Any suggestions to further debug this will be great.

**Comment by ThiloteE** [comment]
Thank you for the report @kaustubhmote. Are you also on gtk 3, or is this on gtk4?

The only thing that has not been tried yet, is to use gtk 4 instead of 3, i think.

Sorry, I personally am out of ideas. Hoping for one of the more experienced maintainers to have a look at this.

**Comment by kaustubhmote** [comment]
No issues. Thanks for your input. This was still GTK3. Will try out GTK4 when I get a chance.

**Comment by ThiloteE** [comment]
If i may ask: How exactly do you two start Jabref?

**Comment by ThiloteE** [comment]
I found some information related to one of the earlier error messages in your log:

```
Jan 12, 2022 8:02:40 AM com.sun.javafx.application.PlatformImpl startup
WARNING: Unsupported JavaFX configuration: classes were loaded from 'module org.jabref.merged.module', isAutomatic: false, isOpen: true
```

https://stackoverflow.com/questions/67854139/javafx-warning-unsupported-javafx-configuration-classes-were-loaded-from-unna

Not sure, if it can help you starting the application properly, or if it is something that would need to be configured within the code of Jabref.

**Comment by ThiloteE** [comment]
Also, was this a clear install of Jabref, or did you have an older version of Jabref installed prior? Maybe a reset of the preferences would help? (Back them up before doing this)

![image](https://user-images.githubusercontent.com/73715071/151705902-b5e5dc6c-49e2-4b8a-9407-7619f8319f1d.png)

**Comment by kaustubhmote** [comment]
I usually use the standalone linux binaries and point a shortcut to `bin/JabRef`. I have tried installing from `deb`, which gives the same results, so does compiling from source using gradle. At some point I thought maybe my database has some errors, but this behavior remains with a test library with a single entry. I tried the preference reset option, but the issue seems to be the same in both X11 and Wayland. Still have not tested on GTK4 though.

**Comment by cherrywoods** [comment]
The first version I tested was installed via snap on a fresh operating system installation. 
There may have been old configuration files flying around, as I copied data from my old PC.
However, at least the theme was not carried over to the new JabRef, so I figure no old configuration files were used.

> I usually use the standalone linux binaries
I used these to test the most recent development build, I didn't clean up the snap installation of JabRef.
When trying other versions I also did not do any cleanup.

**Comment by ThiloteE** [comment]
@Siedlerchr would you be able to confirm that JabRefs JavaFX is resolved as named modules? In other words, it is loaded from the module- path and not from the class-path, as per https://stackoverflow.com/questions/67854139/javafx-warning-unsupported-javafx-configuration-classes-were-loaded-from-unna ?

**Comment by Siedlerchr** [comment]
You can ignore the message by JavaFX named modules. It's just a warning and comes by the fact that we merge JavaFX modules and other not yet modularized libraries together into a new module (fairly complex topic) which has a different name. This also does not have an impact. 
 
A typical memory impact: Try to disable autocomplete (if enabled) It has been known to affect the performance. 
The most recent dev build (should be also snap edge) also has an option to disable indexing of PDFs for the full text search. Maybe you can try to test this.

**Comment by ThiloteE** [comment]
Thank you!

**Comment by cherrywoods** [comment]
I just tested the latest dev build with the suggested configuration. 
The issue is not resolved. Memory consumption is fine on startup and after loading a .bib file. It starts to blow up once I open any entry of the .bib file. 

Here is the log, in case this is interesting:
```
Feb 03, 2022 9:54:51 AM com.sun.javafx.application.PlatformImpl startup
WARNING: Unsupported JavaFX configuration: classes were loaded from 'module org.jabref.merged.module', isAutomatic: false, isOpen: true
2022-02-03 09:54:52 [JavaFX Application Thread] org.jabref.logic.net.URLDownload.bypassSSLVerification()
WARN: Fix SSL exceptions by accepting ALL certificates
2022-02-03 09:54:53 [JavaFX Application Thread] org.jabref.gui.theme.ThemeManager.updateThemeSettings()
INFO: Not updating theme because it hasn't changed
2022-02-03 09:54:53 [JavaFX Application Thread] org.jabref.gui.theme.ThemeManager.updateThemeSettings()
INFO: Theme set to Theme{type=EMBEDDED, name='Dark.css'} with base css StyleSheet{jrt:/org.jabref/org/jabref/gui/Base.css}

(JabRef:4963): Gdk-WARNING **: 09:54:55.263: XSetErrorHandler() called with a GDK error trap pushed. Don't do that.
2022-02-03 09:56:28 [JavaFX Application Thread] org.jabref.gui.JabRefDialogService.notify()
INFO: Preferences recorded.
2022-02-03 09:57:18 [pool-2-thread-2] org.jabref.gui.JabRefDialogService.notify()
INFO: Opening: '/SOMEPATH ... /library.bib'
```

**Comment by ThiloteE** [comment]
So this was on gtk4, with indexing disabled, autocomplete disabled and with x11?

**Comment by cherrywoods** [comment]
Sorry, no. This was with gtk3 and wayland, but indexing and autocomplete disabled. 
So, this was in the setup where I first encountered the issue, the machine I am normally working with.

**Comment by ThiloteE** [comment]
how can i reproduce the measurements mentioned in https://github.com/JabRef/jabref/issues/8416#issuecomment-1019164316 ? Which program or commands did you use?

**Comment by kaustubhmote** [comment]
Before starting, I started logging the memory usage for Jabref on a separate terminal using a bash script that saves the output from `ps` every second

```shell
# this is logging.sh
#!/usr/bin/bash
echo "USER PID %CPU %MEM VSZ RSS TTY STAT START TIME COMMAND"
while true; do
    echo "$(date '+%Y-%m-%d %H:%M:%S') $(ps aux | grep "[J]"abRef)"
    sleep 1
done
```
```bash
$ ./logging.sh > jabref.log &
$ # start Jabref
```

The virtual memory column (VSZ) in `jabref.log` is plotted using matploltib.

**Comment by cherrywoods** [comment]
I just tested with version 5.3 of Jabref. This version also has the issue, but memory consumption is growing much more slowly. So slowly that it's almost usable. 

Newer versions of JabRef still explode memory consumption as soon as I open an entry on my machine.

**Comment by ThiloteE** [comment]
What I do not understand is why a lot of Linux users have problems with Jabref and seemingly JDK. Jabref comes with its own version of jdk and javafx, right @Siedlerchr ? Why do people still have issues starting Jabref then, if it should use its own version of these programs?

**Comment by cherrywoods** [comment]
I will test gtk4 in a virtual machine this weekend

**Comment by ThiloteE** [comment]
@HoussemNasri Sorry to pull you into this conversation, but since you seem to be knowledgable about performance issues, maybe you have an idea, how we could debug this?

**Comment by HoussemNasri** [comment]
I'm unable to reproduce this on a virtual machine running Ubuntu 20.04. However, as @kaustubhmote mentioned, when selecting an entry or creating a new one, there is a significant increase in virtual memory usage; though RAM usage was normal. The interesting part is that both actions create a WebView instance and initialize the preview panel, which is what caused the freeze in #8446. So I looked for a connection between *virtual memory*, *webviews*, and *Linux* but didn't find much; I found these two Stackoverflow posts and an old JDK that was fixed in JDK 9: 
- [Virtual Memory Usage from Java under Linux, too much memory used](https://stackoverflow.com/questions/561245/virtual-memory-usage-from-java-under-linux-too-much-memory-used)
-  [Exploding memory usage in Javafx WebView while loading many pages](https://stackoverflow.com/questions/36266013/exploding-memory-usage-in-javafx-webview-while-loading-many-pages) 
- [JDK bug](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8159594).

Further actions that result in the creation of a WebView and might trigger this issue:
- Opening Preferences and selecting the *Entry preview* tab.
- Opening the global search results window.
- Selecting an *openoffice/libreoffice* style.

**Comment by ThiloteE** [comment]
All open / in progress memory leak bugs in jdk since 2016:

https://bugs.openjdk.java.net/browse/JDK-8277848?jql=project%20%3D%20JDK%20AND%20status%20in%20(Open%2C%20%22In%20Progress%22)%20AND%20text%20~%20%22leak%22%20AND%20created%20%3E%3D%202016-04-02%20AND%20created%20%3C%3D%202022-04-02%20ORDER%20BY%20affectedVersion%20DESC

**Comment by cherrywoods** [comment]
I also couldn't recreate the issue in a virtual machine. I will test whether the three other actions that create a WebView trigger the memory issue in the instance where I experience the problem.

**Comment by cherrywoods** [comment]
I now checked on the machine where I experience the issues:
 - merely opening preferences induces a similar increase in memory as opening an entry
 - opening the Entry preview tab does not introduce further increase in memory usage
 - neither does selecting an openoffice/libreoffice style

The issue has changed quite a bit, since I opened this issue. Now it is mainly that memory usage keeps increasing once starting JabRef at a slow pace. Pace increases when opening an entry or the preferences, but I can keep it running for 10 minutes or so until the memory usage gets problematic.

**Comment by HoussemNasri** [comment]
@cherrywoods Can you possibly download JProfiler to profile the memory usage of JabRef? It will undoubtedly help us in identifying the issue. JProfiler is not free, but it does come with a 10-day trial period when you use it for the first time. The installation instructions for Linux may be found [here](https://www.ej-technologies.com/resources/jprofiler/help/doc/main/installing.html). Following installation, you can follow these steps to generate an HTML report; I've also attached a video for your convenience:
1. Start JabRef.
2. Open Jprofiler and attach it to the JabRef process.
3. Select Live Memory.
4. Filter classes using this filter `org.jabref.*,?.javafx.?`.
5. Sort the results by size.
6. Use JabRef until you notice extensive memory usage.
7. Export the results in HTML format.


https://user-images.githubusercontent.com/21198231/161450042-c25646ef-64cf-47d3-b587-4b72bfe26d85.mp4

**Comment by cherrywoods** [comment]
Thanks for looking into this. I have profiled a session of using JabRef. 

As I mentioned the issue has changed significantly since I opened this issue. The memory consumption now stays tolerable for a fair amount of time. I was using JabRef as I usually would now for some time (maybe 10 min). According to System Monitor it consumed 1.1GB memory in the end. From my perspective, this seems to be too much, but maybe this number is expected?

I made three reports, which are included in the zip file below. The first is when starting jabref (before opening bib file) as a baseline, then after opening the bib file (for reference the bib file has a little less than 150 entries) and then after the jump in memory after opening the first entry. There System Monitor reports  about 900 MB memory usage, which goes up to 1.1 GB after some time (file Final_...)

Short Note: I thought first the report would contain memory data over time, so first I just exported at the end of my session of using JabRef. This is the file Final_... contained in the zip archive. Then I noticed that it only contains a snapshot and opened JabRef a second time to create the further three reports. 
I didn't want to spend more time on this for now, but if the above presents a problem, I can do it properly once I use JabRef the next time. 

[JabRef_Linux_2022_04_04.zip](https://github.com/JabRef/jabref/files/8409960/JabRef_Linux_2022_04_04.zip)

**Comment by cherrywoods** [comment]
I checked the memory consumption on another Linux machine and it is comparably high. I guess the issue on my other machine has fixed itself then and the behavior is normal...

**Comment by ThiloteE** [comment]
Great that it seems to work for you now cherrywoods. Yes, 1.1 gigabyte of RAM usage is pretty normal for JabRef nowadays. I get similar measurements with Windows Task Manager on my machine. Yes, it is very high in comparison to other programs and the lower the hardware requirements, the better. 

I should have asked earlier: How much RAM does your machine have available in total?

Having 1.1gb of JabRef in memory does not really support the argument that what you encountered is a memory leak and something is out of the ordinary. What kaustubhmote reports might, though. So I would rather have this issue open, until kaustubhmote's issue is also solved.

@kaustubhmote what about you? https://github.com/JabRef/jabref/issues/8416#issuecomment-1019164316

**Comment by cherrywoods** [comment]
My machine has 8GB of RAM. It's too bad that I can not reproduce the issue anymore that I had initially. That was using up all my memory and rising continuously. 

My impression is that memory still keeps rising over time, which hints at a memory leak, although not a drastic one. Maybe the reports I generated bear some trace of it.

**Comment by ThiloteE** [comment]
Alright. Thanks. Please, which Jabref version did you use for the measurments in https://github.com/JabRef/jabref/issues/8416#issuecomment-1087541116

You can find this information within Jabref at `Help>About Jabref > `

**Comment by cherrywoods** [comment]
5.5.6 installed via snap

**Comment by kaustubhmote** [comment]
I also cannot seem to reproduce this issue anymore using JabRef versions 5.2 to 5.6-dev. JabRef seems to be stable, and consuming approx 1.1 GB memory (Virtual Memory Usage is still high at ~112 GB, but seems to be stable on both X11 and Wayland. I have not had a crash running JabRef in the background all day long, with some intermittent interaction. Over time, the memory usage does increase. I reached something like 2.5 GB after several hours. This does not seem to go back to 1.1 even if all the libraries are closed. Perhaps this depends on something external that meanwhile got fixed?

**Comment by ThiloteE** [comment]
It is already good that there does not seem to be a crash any more! Nice!


> JabRef versions 5.2 to 5.6-dev. JabRef seems to be stable, and consuming approx 1.1 GB memory (Virtual Memory Usage is still high at ~112 GB, but seems to be stable on both X11 and Wayland.

~ 112 GB seems quite a lot, though. My machine would not have enough free space!

My current thought about this is to have a look into the virtual memory to check which files are in there exactly. On windows 10, there is the "[performance monitor](https://www.windowscentral.com/how-use-performance-monitor-windows-10)" tool available. I also found [VMMap](https://docs.microsoft.com/en-gb/sysinternals/downloads/vmmap) that apparently can analyse (virtual) memory usage patterns.

For Java applications (such as Jabref) in general, I found this beginners guide to analyze performance issues (from 2016, but seems well explained) https://www.vogella.com/tutorials/JavaPerformance/article.html. Here are some tools mentioned for Linux too. Maybe that will help in diagnosing this further.

**Comment by ThiloteE** [comment]
I tried the newest development version.

JabRef 5.6--2022-04-11--1c782b7
Windows 10 10.0 amd64 
Java 17.0.2 
JavaFX 18+12

Next to others, these are notable changes that probably had an effect on RAM usage:

- pdfbox (version 3.0) https://github.com/JabRef/jabref/pull/8658
- https://github.com/JabRef/jabref/pull/8624

Now on my Windows machine with a library of 99 entries and ~ 90 linked pdf files, I see:


|Variables||
|----|----|
| After the start of JabRef | 600 - 800 MB RAM. If I do NOTHING and just let JabRef calm down, it usually drops to ~600 (even saw 470 once!), but NOT ALWAYS. Once, I saw stabilization at 750 and only when I opened the entry editor, it started to drop. Maybe it was a coincidence.|
| Open the preferences | It jumps to something in between 750 - 1000 MB. |
| Close the preferences | After a short while it drops again, but stabilizes at a higher level. Usually something in between 670-1000 MB.|
| Open the "entry editor" |  In version 5.5 this led to an INCREASE of the RAM usage (~ +200MB). Now, it leads to a DECREASE or seemingly has no effect on RAM. I saw it dropping to ~ 600 twice. At other times it was already at ~ 600 and there was barely any change. |
| Write XMP metadata to linked PDF files | RAM usage jumps to ~ 2000 MB. It does NOT drop back down until I restart JabRef. (I waited ~ 10-15 minutes until I restarted JabRef) |

I am curious about
- the size of your libraries
- and if you experience similar behaviours at your Linux machines?

We are thinking of releasing a new version very soon, so it would be nice to receive feedback. Hopefully all is alright!

**Comment by cherrywoods** [comment]
I also tested it out:

JabRef 5.6--2022-04-12--d50adc1
Linux 5.13.0-39-generic amd64 
Java 17.0.2 
JavaFX 18+12

 - My Library has 165 entries
 - On Start: 720 MiB of Memory according to System Monitor
 - Increase to 738 MiB when opening the Preferences (no notable increase)
 - No decrease after closing the Preferences
 - Opening an Entry: Jump to 752 MiB
 - Increase to 766 MiB after opening further entries
 - Increase continues to 788 MiB while waiting for some time

So, very different behaviour. Much better though!

**Comment by ThiloteE** [comment]
meta-issue: #8416

**Comment by ThiloteE** [comment]
Closing this, as it was reported there are no issues anymore (except very high Swap space usage)

Recommendations to reduce high swap space usage:
- If JabRef was compiled from source: check if correct up to date java and javafx was used to compile (e.g. currently jfx 18.01)
- If JabRef was used from commandline: use  `\JabRef\lib\runtime\bin\JabRef`. Do NOT use `/bin/JabRef`. Follow [CLI documentation](https://docs.jabref.org/advanced/commandline)
- Avoid known causes for high CPU/RAM. Currently these causes are to enable the following preferences or use following features: Field formatters (normalize date), autosave, autocomplete, search index for linked PDFs.
- A special case is "writing XMP metadata to linked PDFs". It causes high RAM usage, which will not go down, but it only requires a restart to reset.
- In general: use smaller bibliographies. The more entries --> the higher the RAM usage. The more entries -->  the higher the CPU usage. The more entries -->  the longer it takes for certain actions to progress.
